### ISSUE TITLE
[RLLib] [Question] Control over the whole training workflow

### ISSUE THREAD

**Issue by ankeshanand** [ray-project/ray#4561]
Hi,

I appreciate how RLLib tries to abstract and encapsulate various different modules in an RL pipeline, but I have been finding it sometimes hinders hacking with the code and implementing non-standard algorithms. Here's something I have been trying to implement (https://papers.nips.cc/paper/7507-meta-gradient-reinforcement-learning): 

![image](https://user-images.githubusercontent.com/2134563/55578819-cb9b5e00-56e4-11e9-94b9-3efd4caaf0fa.png)

Implementing something like this requires not only changing the policy graph but also modifying how we sample and put items in a replay buffer. Is there an example about how to go about a custom workflow like this?

**Comment by ericl** [comment]
@ankeshanand thanks for bringing this up. I think the way to go in these sort of research use cases is to bypass the built in "Agent" (soon renamed to Trainer) classes. Instead, use the policy graph and policy evaluation classes directly with Ray to implement your training workflow.

For example, suppose you wanted to collect two sets of trajectories T, T', and then optimize theta, n, and repeat. You could do this by creating a custom policy graph class, and a set of policy evaluation workers for parallel sample collection:

```
... define your own env and policy graph classes classes ...
policy = ...
n = ...
workers = [PolicyEvaluator.as_remote().remote(env_creator, policy_graph)
    for i in range(10)]

while True:
  T1 = SampleBatch.concat(ray.get([w.sample.remote() for w in workers]))

  # can adjust policy params if you want
  for w in workers:
     w.for_policy.remote(lambda policy: policy.do_some_custom_adjustment(n))
  T2 = SampleBatch.concat(ray.get([w.sample.remote() for w in workers]))

  policy.learn_on_batch(T1)
  n = update_meta_param(T2)
```

It could also be possible to coordinate the same workflow with some existing Agent class and custom callbacks, but you get much more control writing your own Agent/Trainer or using the lower level RLlib components directly. Let me know if this makes sense.

**Comment by ankeshanand** [comment]
Gotcha, thanks, I will try this! And yes, the name Trainer makes more sense! 

Aside: One of the things I found confusing that `self.optimizer.step()` does not not only optimize the policy graph with a batch (which is what I would have guessed intuitively, but also co-ordinates collecting the samples too). 

Relatedly as a new user, I think it would have easier for me to onboard if things were a lot more explicit (even if it means I write a few more lines of code), and not hidden behind lots of abstractions.

**Comment by ericl** [comment]
That's a good point! Maybe optimizer isn't the right term either, since sample collection is a major part of it's functionality.

**Comment by ericl** [comment]
Btw, cleaned this up and made a runnable example here: https://github.com/ray-project/ray/pull/4554/files#diff-b58a7ecdf4c8cdfe99316a5de0893f58R64
